samples/gui_test_wgpu/src/gui_test_wgpu.zig

```diff
@@ -579,40 +585,65 @@ fn draw(demo: *DemoState) void {
+        std.log.err("Failed to initialize GLFW library: {}", .{err});
         return;
     };
     defer zglfw.terminate();
 
-    // Change current working directory to where the executable is located.
-    {
+    if (!zems.is_emscripten) {
+        // Change current working directory to where the executable is located.
         var buffer: [1024]u8 = undefined;
         const path = std.fs.selfExeDirPath(buffer[0..]) catch ".";
         std.os.chdir(path) catch {};
     }
 
-    const window = zglfw.Window.create(1600, 1000, window_title, null) catch {
-        std.log.err("Failed to create demo window.", .{});
+    if (zems.is_emscripten) zglfw.WindowHint.set(.client_api, @enumToInt(zglfw.ClientApi.no_api));
+    const window = zglfw.Window.create(1600, 1000, window_title, null) catch |err| {
+        std.log.err("Failed to create demo window: {}", .{err});
         return;
     };
     defer window.destroy();
     window.setSizeLimits(400, 400, -1, -1);
 
-    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
-    defer _ = gpa.deinit();
+    var gpa = if (zems.is_emscripten) zems.EmmalocAllocator{} else std.heap.GeneralPurposeAllocator(.{}){};
+    defer _ = if (!zems.is_emscripten) gpa.deinit();
 
     const allocator = gpa.allocator();
 
-    const demo = create(allocator, window) catch {
-        std.log.err("Failed to initialize the demo.", .{});
+    const demo = create(allocator, window) catch |err| {
+        std.log.err("Failed to initialize the demo: {}", .{err});
         return;
     };
     defer destroy(allocator, demo);
 
-    while (!window.shouldClose() and window.getKey(.escape) != .press) {
-        zglfw.pollEvents();
-        try update(demo);
-        draw(demo);
+    if (zems.is_emscripten) {
+        // requestAnimationFrame usage is needed to avoid flickering, otherwise could work in loop as well
+        zems.requestAnimationFrameLoop(&tickEmcripten, demo);
+        // TODO: ideally we would store all state in globals and return, but for now I don't want to refactor it, so lets spin forever
+        while (true) zems.sleep(1000);
+    } else while (!window.shouldClose() and window.getKey(.escape) != .press) {
+        try tick(demo);
     }
 }
+
+pub fn tick(demo: *DemoState) !void {
+    zglfw.pollEvents();
+    try update(demo);
+    draw(demo);
+}
+
+usingnamespace if (zems.is_emscripten) struct {
+    pub export fn tickEmcripten(time: f64, user_data: ?*anyopaque) callconv(.C) zems.EmBool {
+        _ = time;
+        const demo = @ptrCast(*DemoState, @alignCast(@alignOf(DemoState), user_data.?));
+        if (demo.gctx.canRender()) tick(demo) catch |err| {
+            std.log.err("animation frame canceled! tick failed with: {}", .{err});
+            return .false;
+        } else {
+            std.log.warn("canRender(): Frame skipped!", .{});
+        }
+        return .true;
+    }
+} else struct {};
+extern fn tickEmcripten(time: f64, user_data: ?*anyopaque) callconv(.C) zems.EmBool;
```

Comment:
I think we should look into abstracting this away in zglfw. But I'm not sure exactly what it would look like.... something like zglfw init sets up a js animation callback which sets some redraw bool that we can check in our application loop. It's not the ideal way of handling browser gpu sync but I think it's "good enough". Tight gpu synchronization requires using a better platform layer than what glfw provides anyway.

It would of course mean than zems becomes an optional dependency of zglfw, which I think is fine.

What do you guys think? @Deins @michal-z

Reply:
I think its not glfws fault. (at least for web) 
Sdl2 had the same issue a while back, it required user to use emscripten specific callbacks for web (at least in older versions that I remember). Random example from web: https://github.com/erik-larsen/emscripten-sdl2-ogles2/blob/master/src/hello_triangle.cpp#L144

Core issue is that on web nothing can block main thread. (and `WebWorker` threads can't access any web api, so its not even worth mentioning). Emscripten provides `asyncify` magic that will try to detect functions that can block and do some crazy stuff to unwind stack etc. to allow blocking code to actually not block. But it has its limitations.
In general to smoothly render there is no alternative as far as I know than using requestAnimationFrame.

I think unless API is ground up built, not to ever block, it doesn't matter, it will require getting rid of main loop and using web specific callback interface. Nice example of how such api might look like is C library [sokol](https://github.com/floooh/sokol) which just tries to provide cross platform window, input etc.
But I think its out of scope of `zig-gamedev` which mostly just provides zig bindings and libraries which you match together as you wish.

So yeah,  currently maybe is good idea just to have one example that show how to compile for web with emscripten.

Reply:
Yeah you're right, the problem is trying to fudge a desktop platform abstraction to work with browsers and they we're never designed to do so.

In my own projects I use a custom platform abstraction with completely separate desktop and web backends. Maybe when it is more mature I will look at contributing the web portion to zig-gamedev. The use-case here I think is being to quickly be able to take something written on-top of glfw and get it going in a browser. Which wouldn't the path I would recommend for any serious project. But it's a usecase that *can* be supported nonetheless.

> So yeah, currently maybe is good idea just to have one example that show how to compile for web with emscripten.

Yeah I think this is the way to go, for now it can just be glfw+webgpu. But in the future we could potentially reuse `zems` to show how to get sdl+opengl in the browser too. Sorry for the distraction by suggesting that we should try to get existing samples to work.

Let's see if @michal-z agrees

Reply:
Yes, I agree. One sample compiled for the web is enough for this PR. It will provide nice example and give us experience/knowledge about what exactly is needed - we can think about better/different abstraction in the future.